packages/workbox-cache-expiration/Plugin.mjs

```diff
@@ -76,6 +79,10 @@ class Plugin {
     this._config = config;
     this._maxAgeSeconds = config.maxAgeSeconds;
     this._cacheExpirations = new Map();
+
+    if (config.purgeOnQuotaError) {
+      registerCallback(() => this.deleteCacheAndMetadata());
```

Comment:
Is the plan that `registerCallback()` might/will get invoked in more places in the future? From what I can tell in this PR, this is the only place it's used, which makes the `quota.mjs` module seem like over-kill for just this.

Reply:
Yes, along the lines of https://github.com/GoogleChrome/workbox/pull/1505#discussion_r194867665

In addition to giving developers a general-purpose way of registering a "cleanup" callback, we may want to make use of it elsewhere in the Workbox codebase—potentially as part of the precaching module, along the lines of #1312.

Reply:
Yeah, I get the intention of the change, I'm just wondering if it could be somehow better integrated into the rest of workbox.

Did you consider creating a new lifecycle callback: maybe something like `cacheWriteDidFail`? And in the same way that we can register a generic fetch failure handler, we could register generic cache failure handler.

Anyway, just a thought, I don't want to block this change if a better solution isn't obvious.

Reply:
I felt like using the lifecycle callback naming conventions would be misleading, since here you're registering a callback that has a "global" (or `workbox.core`) scope, which will be invoked regardless of which strategy triggers the quota exception. That doesn't match the behavior of the lifecycle callbacks, which are all scoped to a given strategy.

`workbox.routing.setCatchHandler()` is closer in semantics, but that only supports setting a single callback, while we need support for multiple quota exception callbacks. To me, `register...` sounds more in line with being able to create multiple callbacks than `set...`, but I could change the naming if you feel strongly.

Reply:
> I felt like using the lifecycle callback naming conventions would be misleading, since here you're registering a callback that has a "global"...

Right, I wasn't suggesting just using the name, I was suggesting we actually add a `cacheWriteDidFail` lifecycle callback. While I understand that it's not global and would have to be added per strategy, that seems like a benefit to me, not a shortcoming. And users could write their strategy code in such a way that sharing the same callbacks across all their strategies isn't so bad.

My concern is that global handlers can lead to unpredictable behavior that's harder to test, especially for folks who are using Workbox in a more modular way, or anyone building a third-party plugin.

Anyway, I don't want to block this feature being added, especially if we know folks are running into quota issues today. And you've obviously thought about the trade-offs, so I'm happy with whatever you think is best.

But I think we should revisit it later. I also think it's probably worth adding a `catchWriteDidFail` lifecycle method at some point regardless, so strategy plugin authors can handle these cases individually without potentially affecting plugins they don't own.

Reply:
I filed https://github.com/GoogleChrome/workbox/issues/1530 to capture that. To address the immediate quota issues, I'd like to go with a global approach.